### PR TITLE
Scripts Integration for Story Mode

### PR DESCRIPTION
**Pull request by Arik**
Tested using my own story.json with a script and story-default.json

Currently the "scripts" field is optional in the story json files. Should that be changed?

Also rlbot_to_player_config() and script_to_script_config() share the same first lines but I'm not sure if those lines should be a function instead.

### CODE REVIEW

**Comment by VirxEC** [comment]
I like this idea

**Comment by azeemba** [comment]
I won't be able to have a look till next week. Sorry about this. 

Do you mind elaborating on what the use case is here?

**Comment by CoolArik1** [comment]
Sure. This change allows scripts to be run in story mode challenges. In a story config file, you can add scripts just like you add bots:
```json
"bots": {},
"scripts": {
    "0g": {
        "path": ["$RLBOTPACKROOT", "RLBotPack", "CustomMutators", "ZeroG", "zero_g.cfg"]
    }
},
```
Then you can add scripts to challenges by using the "scripts" field:
```json
{
    "id": "INTRO-1",
    "humanTeamSize": 1,
    "opponentBots": ["sniper"],
    "scripts": ["0g"],
    "map": "BeckwithPark_Midnight",
    "display": "There's no gravity!"
}
```
Then when using that story config file, the Zero G Mutator script runs during the first intro challenge.

File: rlbot_gui/story/story_challenge_setup.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
     PlayerConfig,
     MatchConfig,
     MutatorConfig,
-    Team,
+    Team, ScriptConfig,
```
Comment (azeemba) [inline rlbot_gui/story/story_challenge_setup.py:28]: Best to maintain the style of surrounding code and add a newline

File: rlbot_gui/story/story_challenge_setup.py [hunk 2]
```diff
@@ -119,6 +120,23 @@ def rlbot_to_player_config(player: dict, team: Team):
     return player_config
 
 
+def script_to_script_config(script: dict):
```
Comment (azeemba) [inline rlbot_gui/story/story_challenge_setup.py:123]: We don't seem to use any other properties in the dict so can we just directly accept the path instead?

File: rlbot_gui/story/story_challenge_setup.py [hunk 3]
```diff
@@ -119,6 +120,23 @@ def rlbot_to_player_config(player: dict, team: Team):
...
+        script_path = path.join(*script_path)
+
+    if "$RLBOTPACKROOT" in script_path:
+        for bot_folder in rlbot_gui.bot_folder_settings["folders"].keys():
+            adjusted_folder = path.join(bot_folder, "RLBotPack-master")
+            subbed_path = script_path.replace("$RLBOTPACKROOT", adjusted_folder)
+            if path.exists(subbed_path):
+                print("it exists! (script)")
+                script_path = subbed_path
+                break
```
Comment (azeemba) [inline rlbot_gui/story/story_challenge_setup.py:135]: Looks like this code is being duplicated from `rlbot_to_player_config`. Can we instead pull this logic out into a function and use it in both places?

File: rlbot_gui/story/story_runner.py [hunk 4]
```diff
@@ -14,7 +14,7 @@
     get_bots_configs,
     get_cities,
     get_story_settings,
-    get_challenges_by_id
+    get_challenges_by_id, get_scripts_configs
```
Comment (azeemba) [inline rlbot_gui/story/story_runner.py:17]: Probably best to add it on a new line

**Review by azeemba — COMMENTED** [review]
Thanks for the PR and sorry for the late response.

The example usage does help me better understand the feature. Do you have any specific plans to use this feature? Or did someone else in Discord mention a use case? Knowing this would help me better make suggestions on the interface and scope of this change.

**Comment by Arik** [comment]
> Do you have any specific plans to use this feature?

I have been working on a story that uses this feature. The usage in my story is very similar to the example usage.

Ultimately, I want to make adding scripts to story challenges similar to adding bots to story challenges.

**Review by azeemba — APPROVED** [review]
Thanks for describing your use case, that does seem very exciting!

The changes look good. Do you mind bumping the version in setup.py similar to https://github.com/RLBot/RLBotGUI/commit/6d2d2bd6591469f0aeb3347c79b474dddbc474ac. This will automatically push out a release when the PR lands.

**Comment by azeemba** [comment]
Thanks again for your contribution!

**Pull request merged**